### PR TITLE
Add dev tooling, platform-aware torch, and experiment monitor

### PR DESCRIPTION
**Pull request by Jason-Adam**
## Summary
- **Dev infrastructure**: ruff, mypy, pytest, pre-commit hooks, and Makefile for standard dev workflows
- **Platform-aware torch**: uv resolves torch from CUDA cu128 index on Linux, PyPI on macOS ARM64
- **Experiment monitor**: lightweight HTML dashboard (monitor.py) on localhost:8080 — parses run.log for live training metrics, results.tsv for experiment history, optional remote GPU status via SSH
- **Schedule extraction**: pure schedule functions moved from train.py to schedules.py for testability

### Review fixes (ed5875d)
- Aligned `_PROGRESS_RE` regex with actual `train.py` log format (parenthesized percent, colon-separated fields)
- Added HTML escaping (`html.escape`) for all file-derived values in dashboard and report to prevent XSS/malformed pages
- Fixed `generate_report()` to use consistent column ordering across entries with different summary keys
- Removed `StrictHostKeyChecking=no` from SSH probe — uses system defaults (`~/.ssh/known_hosts`)
- Added `warmdown_ratio <= 0` guard in `get_lr_multiplier()` to prevent `ZeroDivisionError`
- Added `ramp_steps <= 0` validation in `get_muon_momentum()` with clear `ValueError`
- Added tests for new edge cases (24 total, all passing)

## Test plan
- [x] `make lint` passes (ruff + mypy)
- [x] `make test` passes (pytest — 24 tests)
- [x] `uv sync` works on both macOS and Linux
- [x] `python monitor.py` serves dashboard on :8080
- [x] Pre-commit hooks run on commit

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: monitor.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+
+# ---------------------------------------------------------------------------
+# Parsers
+# ---------------------------------------------------------------------------
+
+# Matches: step  142 | 47.3% | loss 3.421 | lr_mult 0.83 | 234.5 ms/step | 12345 tok/sec | mfu 42.1% | epoch 1 | 158.3s remaining
+_PROGRESS_RE = re.compile(
+    r"step\s+(\d+)\s*\|\s*([\d.]+)%\s*\|\s*loss\s+([\d.]+)\s*\|\s*lr_mult\s+([\d.]+)"
+    r"\s*\|\s*([\d.]+)\s*ms/step\s*\|\s*([\d.]+)\s*tok/sec\s*\|\s*mfu\s*([\d.]+)%"
+    r"\s*\|\s*epoch\s+(\d+)\s*\|\s*([\d.]+s\s*remaining)"
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply monitor.py]: Fixed in ed5875d. Aligned the regex with the actual `train.py` format on line 782: parenthesized percent `(47.3%)`, colon-separated fields (`loss:`, `lrm:`, `dt:`, `tok/sec:`), and comma-formatted tok/sec.

File: monitor.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+            num = len(state.history) - i
+            rows_html += (
+                f"<tr>"
+                f"<td>{num}</td>"
+                f"<td>{row.get('commit', '')[:7]}</td>"
+                f"<td>{row.get('val_bpb', '')}</td>"
+                f"<td>{row.get('memory_gb', '')}</td>"
+                f'<td style="color:{color}">{st}</td>'
+                f"<td>{row.get('description', '')}</td>"
+                f"</tr>"
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply monitor.py]: Fixed in ed5875d. All file-derived values now go through `_esc()` (`html.escape`) in both the history panel and report generator.

File: monitor.py [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+    else:
+        history_panel = '<div class="panel"><h2>Experiment History</h2><p style="color:#555">No results.tsv found.</p></div>'
+
+    # --- Summary panel ---
+    if run.summary:
+        items = "".join(
+            f'<div class="summary-item"><div class="summary-key">{k}</div>'
+            f'<div class="summary-val">{v}</div></div>'
+            for k, v in run.summary.items()
+        )
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply monitor.py:418]: Fixed in ed5875d alongside the history panel — summary keys and values now go through `_esc()`.

File: monitor.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+    for e in reversed(entries):
+        ts = time.strftime("%Y-%m-%d %H:%M", time.localtime(e.get("timestamp", 0)))
+        commit = e.get("commit", "")[:7]
+        summary = e.get("summary", {})
+        cells = " ".join(f"<td>{v}</td>" for v in summary.values())
+        rows_html += f"<tr><td>{ts}</td><td>{commit}</td>{cells}</tr>"
+
+    if entries:
+        sample = entries[-1].get("summary", {})
+        headers = " ".join(f"<th>{k}</th>" for k in sample.keys())
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply monitor.py]: Fixed in ed5875d. `generate_report()` now builds a unified `summary_keys` list from all entries (preserving insertion order), then renders each row by key lookup with empty-string fallback for missing keys.

File: monitor.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        "ssh",
+        "-i",
+        key,
+        "-o",
+        "StrictHostKeyChecking=no",
+        "-o",
+        "ConnectTimeout=5",
+        f"{config.user}@{config.host}",
+        f"nvidia-smi --query-gpu={_NVIDIA_SMI_FIELDS} --format=csv,noheader,nounits",
+    ]
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply monitor.py:212]: Fixed in ed5875d. Removed `StrictHostKeyChecking=no` — SSH now uses system defaults (honors `~/.ssh/known_hosts`).

File: schedules.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    warmup_ratio: float,
+    warmdown_ratio: float,
+    final_lr_frac: float,
+) -> float:
+    """Three-phase LR schedule: warmup -> flat -> cooldown."""
+    if progress < warmup_ratio:
+        return progress / warmup_ratio if warmup_ratio > 0 else 1.0
+    elif progress < 1.0 - warmdown_ratio:
+        return 1.0
+    else:
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply schedules.py:23]: Fixed in ed5875d. Added guard: when `warmdown_ratio <= 0`, returns `1.0` until `progress >= 1.0`, then `final_lr_frac`. Test added.

File: schedules.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        cooldown = (1.0 - progress) / warmdown_ratio
+        return cooldown * 1.0 + (1 - cooldown) * final_lr_frac
+
+
+def get_muon_momentum(step: int, ramp_steps: int = 300) -> float:
+    """Momentum ramp from 0.85 to 0.95 over ramp_steps."""
```
[the remark this answers is not in the document]
Comment (Jason-Adam) [reply schedules.py:31]: Fixed in ed5875d. `get_muon_momentum()` now raises `ValueError` for `ramp_steps <= 0`. Test added.
